core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -4175,98 +4179,491 @@ private RelNode convertUpdate(SqlUpdate call) {
         rexNodeSourceExpressionListBuilder.build(), false);
   }
 
-  private RelNode convertMerge(SqlMerge call) {
-    RelOptTable targetTable = getTargetTable(call);
 
-    // convert update column list from SqlIdentifier to String
-    final List<String> targetColumnNameList = new ArrayList<>();
-    final RelDataType targetRowType = targetTable.getRowType();
-    SqlUpdate updateCall = call.getUpdateCall();
-    if (updateCall != null) {
-      for (SqlNode targetColumn : updateCall.getTargetColumnList()) {
-        SqlIdentifier id = (SqlIdentifier) targetColumn;
-        RelDataTypeField field =
-            SqlValidatorUtil.getTargetField(
-                targetRowType, typeFactory, id, catalogReader, targetTable);
-        assert field != null : "column " + id.toString() + " not found";
-        targetColumnNameList.add(field.getName());
+  private Pair<Pair<List<RexNode>, List<List<RexNode>>>, Integer> extractMatchExprs(
+      SqlNodeList updateCallList,
+      LogicalProject mergeSourceRel, Integer nSourceFields,
+      Integer nDestFields, RelOptTable destTable) {
+
+    // Extract the matched clause expressions from the converted merge source select.
+    // These rexnodes will be used to create the final relnode input to the resulting
+    // LogicalTableModify
+    // For the formatting of the merge source select, see rewriteMerge or convertMerge.
+
+    // Get the offset from the start of the select list, to the start of the matched expressions
+    // Note that we add +1 to account for the addition boolean flag we added, which is used
+    // to track if a row matches or doesn't match.
+    Integer totalOffset = nSourceFields + nDestFields + 1;
+
+    // Convert the identifiers in each of the update's target column list into a map of
+    // strings -> idx. This map helps us to convert the orderings during the later iteration.
+
+    // For updates, the values are stored into the merge source select in the order that they
+    // appear in the update's targetColumnList. However, the order that the values appears in the
+    // targetColumnList may not be the same as the order of the columns in the table.
+    // For example:
+    // dest has columns A, B, C
+    // source has columns X, Y, Z
+    //
+    // MERGE INTO dest.. UPDATE SET C = Z, B = Y
+    // results in a targetColumnList of [C, B]
+    // and a mergeSourceRel of Select (sourceCols) (DestCols) (flag), Z, Y
+    //
+    // In this case, we would want the Update rows to be ordered in the same way that they
```

Comment:
The hashmap is a map of SqlUpdate ---> (map of string colname to integer offset), so the ordering will be consistent per each update. 

The we need to have the ordering correct in order to construct the case statement correctly, otherwise we'll have the wrong values in the wrong columns. I've updated the comment with a full explanation, which hopefully makes what I'm doing a bit clearer.

Reply:
This + the updated comment is very clear. Thank you.